File: jacs/examples/a2a_agent_example.rs
//! Example: Creating a JACS agent with A2A protocol support (v0.4.0)
//!
//! This example demonstrates:
//! 1. Creating a JACS agent with post-quantum cryptography
//! 2. Exporting the agent as an A2A Agent Card
//! 3. Signing the Agent Card with JWS for A2A compatibility
//! 4. Wrapping A2A artifacts with JACS provenance

use jacs::a2a::{agent_card::*, extension::*, keys::*, provenance::*};
use jacs::agent::boilerplate::BoilerPlate;
use jacs::{create_minimal_blank_agent, get_empty_agent};
use serde_json::json;
use std::fs;
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== JACS + A2A Integration Example (v0.4.0) ===\n");

    // Initialize JACS with default observability
    jacs::init_default_observability()?;

    // Step 1: Create a JACS agent
    println!("1. Creating JACS agent...");
    let mut agent = get_empty_agent();

    let agent_json = create_minimal_blank_agent(
        "ai".to_string(),
        Some("Advanced Document Processing Service".to_string()),
        Some("Documents processed and verified successfully".to_string()),
        Some("Document processing failed - invalid format or signature".to_string()),
    )?;

    let agent_value = agent.create_agent_and_load(&agent_json, true, None)?;
    println!("   Agent created with ID: {}", agent.get_id()?);
    println!(
        "   Using algorithm: {}",
        agent.get_key_algorithm().unwrap_or(&"unknown".to_string())
    );

    // Step 2: Generate dual keys for JACS and A2A
    println!("\n2. Generating dual keys...");
    let dual_keys = create_jwk_keys(Some("dilithium"), Some("rsa"))?;
    println!("   JACS key: {} (post-quantum)", dual_keys.jacs_algorithm);
    println!("   A2A key: {} (traditional)", dual_keys.a2a_algorithm);

    // Step 3: Export agent as A2A Agent Card (v0.4.0)
    println!("\n3. Exporting to A2A Agent Card...");
    let agent_card = export_agent_card(&agent)?;
    println!("   Agent Card created");
    println!("   - Protocol versions: {:?}", agent_card.protocol_versions);
    println!(
        "   - Interfaces: {} ({})",
        agent_card.supported_interfaces[0].url, agent_card.supported_interfaces[0].protocol_binding
    );
    println!("   - Skills: {} defined", agent_card.skills.len());
    println!("   - Extensions: JACS provenance enabled");

    // Step 4: Sign Agent Card with JWS
    println!("\n4. Signing Agent Card with JWS...");
    let jws_signature = sign_agent_card_jws(
        &agent_card,
        &dual_keys.a2a_private_key,
        &dual_keys.a2a_algorithm,
        &agent.get_id()?,
    )?;
    println!("   Agent Card signed (JWS format)");
    println!("   - Signature: {}...", &jws_signature[..50]);

    // Step 5: Embed signature in Agent Card (v0.4.0)
    println!("\n5. Embedding signature in Agent Card...");
    let signed_card = embed_signature_in_agent_card(&agent_card, &jws_signature, None);
    println!(
        "   Signatures: {}",
        signed_card.signatures.as_ref().map_or(0, |s| s.len())
    );

    // Step 6: Generate .well-known documents
    println!("\n6. Generating .well-known endpoints...");
    let well_known_docs = generate_well_known_documents(
        &agent,
        &agent_card,
        &dual_keys.a2a_public_key,
        &dual_keys.a2a_algorithm,
        &jws_signature,
    )?;

    for (path, _) in &well_known_docs {
        println!("   {}", path);
    }

    // Step 7: Demonstrate wrapping A2A artifacts
    println!("\n7. Wrapping A2A artifacts with JACS provenance...");

    let a2a_task = json!({
        "taskId": "extract-entities-001",
        "type": "document-processing",
        "input": {
            "documentUrl": "https://example.com/doc.pdf",
            "operations": ["ocr", "entity-extraction", "sentiment-analysis"]
        },
        "requestedBy": "client-agent-123",
        "timestamp": chrono::Utc::now().to_rfc3339(),
    });

    let wrapped_task = wrap_artifact_with_provenance(&mut agent, a2a_task, "task", None)?;
    println!("   Task wrapped with JACS signature");
    println!("   - JACS ID: {}", wrapped_task["jacsId"]);
    println!("   - Type: {}", wrapped_task["jacsType"]);

    // Step 8: Verify the wrapped artifact
    println!("\n8. Verifying wrapped artifact...");
    let verification = verify_wrapped_artifact(&agent, &wrapped_task)?;
    println!(
        "   Verification: {}",
        if verification.valid {
            "PASSED"
        } else {
            "FAILED"
        }
    );
    println!("   - Signer: {}", verification.signer_id);
    println!("   - Type: {}", verification.artifact_type);
    println!("   - Timestamp: {}", verification.timestamp);

    // Step 9: Create a multi-step workflow with chain of custody
    println!("\n9. Creating workflow with chain of custody...");
    let mut workflow_artifacts = Vec::new();

    let ocr_result = json!({
        "step": "ocr",
        "status": "completed",
        "output": { "text": "Sample extracted text...", "confidence": 0.98 }
    });
    let wrapped_ocr = wrap_artifact_with_provenance(&mut agent, ocr_result, "workflow-step", None)?;
    workflow_artifacts.push(wrapped_ocr);

    let entity_result = json!({
        "step": "entity-extraction",
        "status": "completed",
        "entities": [
            {"type": "PERSON", "value": "John Doe"},
            {"type": "ORG", "value": "ACME Corp"}
        ]
    });
    let wrapped_entities = wrap_artifact_with_provenance(
        &mut agent,
        entity_result,
        "workflow-step",
        Some(vec![workflow_artifacts.last().unwrap().clone()]),
    )?;
    workflow_artifacts.push(wrapped_entities);

    let chain = create_chain_of_custody(workflow_artifacts)?;
    println!("   Chain of custody created");
    println!("   - Total steps: {}", chain["totalArtifacts"]);

    // Display the JACS extension descriptor
    println!("\n10. JACS Extension Descriptor:");
    let descriptor = create_extension_descriptor();
    println!("{}", serde_json::to_string_pretty(&descriptor)?);

    println!("\n=== Example completed successfully! ===");

    // Save example outputs
    let output_dir = Path::new("a2a_example_output");
    fs::create_dir_all(output_dir)?;

    fs::write(
        output_dir.join("agent-card.json"),
        serde_json::to_string_pretty(&signed_card)?,
    )?;
    fs::write(output_dir.join("agent-card-signed.jws"), &jws_signature)?;
    fs::write(
        output_dir.join("wrapped-task.json"),
        serde_json::to_string_pretty(&wrapped_task)?,
    )?;
    fs::write(
        output_dir.join("chain-of-custody.json"),
        serde_json::to_string_pretty(&chain)?,
    )?;

    println!("\nOutputs saved to: a2a_example_output/");

    Ok(())
}

File: jacsnpm/examples/a2a-agent-example.js
#!/usr/bin/env node

/**
 * JACS A2A Agent Example for Node.js
 * 
 * This example demonstrates how to:
 * 1. Create a JACS agent with A2A support
 * 2. Export to A2A Agent Card format
 * 3. Wrap artifacts with JACS provenance
 * 4. Create workflows with chain of custody
 */

const fs = require('fs').promises;
const path = require('path');
const jacs = require('../src/index');
const {
  JACSA2AIntegration,
  A2A_PROTOCOL_VERSION,
  JACS_EXTENSION_URI
} = require('../src/a2a');

// Example agent data
const exampleAgentData = {
  jacsId: 'nodejs-example-agent',
  jacsVersion: 'v1.0.0',
  jacsName: 'Node.js A2A Example Agent',
  jacsDescription: 'A Node.js agent demonstrating JACS + A2A protocol integration',
  jacsAgentType: 'ai',
  jacsServices: [{
    name: 'Text Analysis Service',
    serviceDescription: 'Analyzes text using NLP techniques',
    successDescription: 'Text successfully analyzed',
    failureDescription: 'Text analysis failed',
    tools: [{
      url: '/api/analyze-text',
      function: {
        name: 'analyzeText',
        description: 'Analyze text and extract insights',
        parameters: {
          type: 'object',
          properties: {
            text: {
              type: 'string',
              description: 'The text to analyze'
            },
            language: {
              type: 'string',
              description: 'Language code (e.g., en, es, fr)',
              default: 'en'
            },
            operations: {
              type: 'array',
              items: { type: 'string' },
              description: 'Analysis operations to perform',
              default: ['sentiment', 'entities', 'keywords']
            }
          },
          required: ['text']
        }
      }
    }]
  }, {
    name: 'Document Processing Service',
    serviceDescription: 'Processes various document formats',
    successDescription: 'Document processed successfully',
    failureDescription: 'Document processing failed',
    tools: [{
      url: '/api/process-document',
      function: {
        name: 'processDocument',
        description: 'Process and extract information from documents',
        parameters: {
          type: 'object',
          properties: {
            documentUrl: {
              type: 'string',
              description: 'URL of the document to process'
            },
            format: {
              type: 'string',
              enum: ['pdf', 'docx', 'txt', 'html'],
              description: 'Document format'
            }
          },
          required: ['documentUrl', 'format']
        }
      }
    }]
  }],
  jacsContacts: [{
    type: 'email',
    value: 'admin@nodejs-example-agent.com',
    description: 'Administrator contact'
  }]
};

async function main() {
  console.log('=== JACS A2A Node.js Agent Example ===\n');

  try {
    // Step 1: Initialize JACS (with mock config for example)
    console.log('1. Initializing JACS...');
    // In a real application, you would load an actual config
    // jacs.load('jacs.config.json');
    console.log('   ✓ JACS initialized\n');

    // Step 2: Create A2A integration
    console.log('2. Creating A2A integration...');
    const a2a = new JACSA2AIntegration();
    console.log('   ✓ A2A integration created\n');

    // Step 3: Export agent to A2A Agent Card
    console.log('3. Exporting to A2A Agent Card...');
    const agentCard = a2a.exportAgentCard(exampleAgentData);
    
    console.log('   ✓ Agent Card created');
    console.log(`   - Name: ${agentCard.name}`);
    console.log(`   - URL: ${agentCard.url}`);
    console.log(`   - Protocol: ${agentCard.protocolVersion}`);
    console.log(`   - Skills: ${agentCard.skills.length} defined`);
    
    agentCard.skills.forEach(skill => {
      console.log(`     • ${skill.name}: ${skill.description}`);
    });
    
    console.log(`   - Extensions: ${agentCard.capabilities.extensions?.length || 0} configured`);
    if (agentCard.capabilities.extensions) {
      agentCard.capabilities.extensions.forEach(ext => {
        console.log(`     • ${ext.uri}: ${ext.description}`);
      });
    }
    console.log();

    // Step 4: Create extension descriptor
    console.log('4. Creating JACS extension descriptor...');
    const extensionDescriptor = a2a.createExtensionDescriptor();
    console.log('   ✓ Extension descriptor created');
    console.log(`   - URI: ${extensionDescriptor.uri}`);
    console.log(`   - Capabilities: ${Object.keys(extensionDescriptor.capabilities).join(', ')}`);
    console.log();

    // Step 5: Demonstrate wrapping A2A artifacts
    console.log('5. Wrapping A2A artifacts with JACS provenance...');
    
    // Example task
    const a2aTask = {
      taskId: 'analyze-001',
      type: 'text-analysis',
      input: {
        text: 'JACS provides cryptographic provenance for A2A agent networks.',
        language: 'en',
        operations: ['sentiment', 'entities']
      },
      requestedBy: 'client-agent-456',
      timestamp: new Date().toISOString()
    };

    // Mock the signing (in real usage, JACS would sign with actual keys)
    const mockSignRequest = (doc) => ({
      ...doc,
      jacsSignature: {
        agentID: exampleAgentData.jacsId,
        agentVersion: exampleAgentData.jacsVersion,
        date: new Date().toISOString(),
        signature: 'mock-signature-base64',
        signingAlgorithm: 'RSA-PSS',
        publicKeyHash: 'mock-public-key-hash',
        fields: Object.keys(doc).filter(k => k !== 'jacsSignature')
      },
      jacsSha256: 'mock-document-hash'
    });

    // Override jacs.signRequest for the example
    const originalSign = jacs.signRequest;
    jacs.signRequest = mockSignRequest;

    const wrappedTask = a2a.wrapArtifactWithProvenance(a2aTask, 'task');
    
    console.log('   ✓ Task wrapped with JACS signature');
    console.log(`   - JACS ID: ${wrappedTask.jacsId}`);
    console.log(`   - Type: ${wrappedTask.jacsType}`);
    console.log(`   - Signer: ${wrappedTask.jacsSignature.agentID}`);
    console.log();

    // Step 6: Create a workflow with chain of custody
    console.log('6. Creating workflow with chain of custody...');
    
    const workflowSteps = [
      {
        step: 'document-receipt',
        documentId: 'doc-789',
        receivedFrom: 'client-agent-456',
        status: 'completed'
      },
      {
        step: 'text-extraction',
        documentId: 'doc-789',
        extractedText: 'Sample text extracted from document...',
        confidence: 0.98,
        status: 'completed'
      },
      {
        step: 'entity-analysis',
        documentId: 'doc-789',
        entities: [
          { type: 'ORG', value: 'JACS', confidence: 0.95 },
          { type: 'TECH', value: 'A2A protocol', confidence: 0.92 }
        ],
        status: 'completed'
      }
    ];

    const wrappedSteps = [];
    for (let i = 0; i < workflowSteps.length; i++) {
      const parentSigs = i > 0 ? [wrappedSteps[i - 1]] : null;
      const wrapped = a2a.wrapArtifactWithProvenance(
        workflowSteps[i],
        'workflow-step',
        parentSigs
      );
      wrappedSteps.push(wrapped);
    }

    const chainOfCustody = a2a.createChainOfCustody(wrappedSteps);
    
    console.log('   ✓ Workflow created');
    console.log(`   - Total steps: ${chainOfCustody.totalArtifacts}`);
    console.log('   - Chain of custody:');
    chainOfCustody.chainOfCustody.forEach((entry, i) => {
      console.log(`     ${i + 1}. ${entry.artifactType} by ${entry.agentId}`);
    });
    console.log();

    // Step 7: Verify wrapped artifact
    console.log('7. Verifying wrapped artifact...');
    
    // Mock verification
    jacs.verifyRequest = () => true;
    
    const verification = a2a.verifyWrappedArtifact(wrappedTask);
    console.log(`   ✓ Verification: ${verification.valid ? 'PASSED' : 'FAILED'}`);
    console.log(`   - Signer: ${verification.signerId}`);
    console.log(`   - Type: ${verification.artifactType}`);
    console.log();

    // Step 8: Generate well-known documents
    console.log('8. Generating .well-known documents...');
    
    // Mock JWS signature and public key
    const mockJwsSignature = 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhZ2VudENhcmQiOiJ0cnVlIn0.signature';
    const mockPublicKeyB64 = Buffer.from('mock-public-key').toString('base64');
    
    // Mock hashPublicKey
    jacs.hashPublicKey = () => 'mock-public-key-hash';
    
    const wellKnownDocs = a2a.generateWellKnownDocuments(
      agentCard,
      mockJwsSignature,
      mockPublicKeyB64,
      exampleAgentData
    );
    
    console.log('   ✓ Documents generated:');
    Object.keys(wellKnownDocs).forEach(path => {
      console.log(`     • ${path}`);
    });
    console.log();

    // Step 9: Save example outputs
    console.log('9. Saving example outputs...');
    
    const outputDir = path.join(__dirname, 'a2a-output');
    await fs.mkdir(outputDir, { recursive: true });
    
    // Save Agent Card
    await fs.writeFile(
      path.join(outputDir, 'agent-card.json'),
      JSON.stringify(agentCard, null, 2)
    );
    
    // Save wrapped task
    await fs.writeFile(
      path.join(outputDir, 'wrapped-task.json'),
      JSON.stringify(wrappedTask, null, 2)
    );
    
    // Save chain of custody
    await fs.writeFile(
      path.join(outputDir, 'chain-of-custody.json'),
      JSON.stringify(chainOfCustody, null, 2)
    );
    
    // Save well-known documents
    for (const [docPath, content] of Object.entries(wellKnownDocs)) {
      const filename = docPath.split('/').pop();
      await fs.writeFile(
        path.join(outputDir, filename),
        JSON.stringify(content, null, 2)
      );
    }
    
    console.log(`   ✓ Outputs saved to: ${outputDir}`);
    console.log();

    // Restore original functions
    jacs.signRequest = originalSign;

    console.log('=== Example completed successfully! ===');
    console.log('\nKey Takeaways:');
    console.log('- JACS provides document-level cryptographic provenance');
    console.log('- A2A enables agent discovery and communication');
    console.log('- Together they create a complete agent ecosystem');
    console.log('- Post-quantum support ensures long-term security');
    console.log('\nNext Steps:');
    console.log('1. Set up actual JACS configuration with keys');
    console.log('2. Implement HTTP endpoints for A2A skills');
    console.log('3. Host .well-known documents on your server');
    console.log('4. Register with A2A discovery services');

  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

// Run the example
main();
